Modules/modFileEncoding.ACM

```diff
@@ -0,0 +1,257 @@
+    ' load the file into the bytsFile array
+    f = FreeFile()
+    Open strFilename For Binary Access Read As #f
+    ReDim byts(LOF(f) - 1)
+    Get #f, , byts()
+    Close #f
+    ReadFromFile = True
+  End If
+End Function
+
+Function WriteToFile(byts() As Byte, strFilename As String) As Boolean
+  ' Writes data from the supplied byte array into the file specified
+  Dim f As Long
+  
+  If FileExists(strFilename) Then Kill strFilename
+  
+  ' Open file ready for output
+  f = FreeFile()
+  Open strFilename For Binary Access Write As #f
+  Put #f, 1, byts
+  Close #f
+  
+  WriteToFile = True
+  
+End Function
+
+Function UCS2_to_UTF8(bytsUCS2() As Byte, ByRef bytsUTF8() As Byte) As Boolean
+  ' This converts a UCS2 Little Endian encoded byte array into UTF8
+  ' Returns True if successful, False if it fails for any reason
+  
+  Dim lenUCS2 As Long
+  Dim b As Long ' The next write position in the UTF8 byte array
+  Dim i As Long ' The current byte position in UCS2 byte array
+  Dim CP_UCS2(1) As Byte
+  Dim CP_UTF8() As Byte
+  
+  UCS2_to_UTF8 = False ' intially set return value to be false
+  
+  lenUCS2 = UBound(bytsUCS2) - LBound(bytsUCS2) + 1
+  
+  ' Check that the length of the UCS2 array is at least 2 bytes
+  If lenUCS2 < 2 Then Exit Function
+  
+  ' Check that the length of the UCS2 array is an even number of bytes
+  If lenUCS2 Mod 2 <> 0 Then Exit Function
+  
+  ' Check that the supplied array has the correct Byte Order Mark (BOM)
+  If bytsUCS2(LBound(bytsUCS2)) <> 255 Or bytsUCS2(LBound(bytsUCS2) + 1) <> 254 Then Exit Function '255 = &HFF, 254 = &HFE
+  
+  ' Redim bytsUTF8 to the maximum it can be ( max 3 bytes per character )
+  ' ignore header of UCS2 (2 bytes) then for each 2 bytes in UCS2 need potentially 3 bytes in UTF8
+  If lenUCS2 = 2 Then
+    ' we need to return an empty array
+    Erase bytsUTF8
+    UCS2_to_UTF8 = True
+    Exit Function
+  End If
+  
+  ReDim bytsUTF8((lenUCS2 - 2) * 3 / 2 - 1)
+  
+  For i = LBound(bytsUCS2) + 2 To UBound(bytsUCS2) Step 2
+    CopyMemory VarPtr(CP_UCS2(0)), VarPtr(bytsUCS2(i)), 2 ' Copy CodePoint into CP array
+    If Not UCS2CP_to_UTF8CP(CP_UCS2, CP_UTF8) Then Exit Function
+    CopyMemory VarPtr(bytsUTF8(b)), VarPtr(CP_UTF8(0)), UBound(CP_UTF8) + 1
+    b = b + UBound(CP_UTF8) + 1
+  Next
+  
+  ReDim Preserve bytsUTF8(b - 1)
+  UCS2_to_UTF8 = True
+  
+End Function
+
+Function UTF8_to_UCS2(bytsUTF8() As Byte, ByRef bytsUCS2() As Byte) As Boolean
+  ' This converts a UTF8 encoded byte array into ucs2 Little Endian
+  ' Returns True if successful, False if it fails for any reason
+  
+  Dim b As Long ' The next write position in the UCS2 array
+  Dim CP_UCS2() As Byte
+  Dim lenUTF8 As Long
+  Dim i As Long
+  Dim j As Long
+  
+  UTF8_to_UCS2 = False ' intially set return value to be false
+  
+  lenUTF8 = UBound(bytsUTF8) - LBound(bytsUTF8) + 1
+  
+  ' Redim bytsUCS2 to the maximum it can be (header plus 2 bytes for each byte in UTF8
+  ReDim bytsUCS2(2 + lenUTF8 * 2)

```

Comment:
Check the BOM is set, as in `UCS2_to_UTF8`

Reply:
Agreed, and done